haskell/src/Lorentz/Contracts/Stablecoin.hs

```diff
@@ -407,15 +398,113 @@ stablecoinContract =
 -- | Parse the metadata registry contract.
 registryContract :: T.Contract (ToT ()) (ToT MetadataRegistryStorage)
 registryContract =
-  $(case readContract @(ToT ()) @(ToT MetadataRegistryStorage) metadataRegistryContractPath $(embedStringFile metadataRegistryContractPath) of
+  $(case readContract' @(ToT ()) @(ToT MetadataRegistryStorage) metadataRegistryContractPath $(embedStringFile metadataRegistryContractPath) of
       Left e -> fail (pretty e)
       Right _ ->
         [|
           -- Note: it's ok to use `error` here, because we just proved that the contract
           -- can be parsed+typechecked.
-          either (error . pretty) snd $
-            readContract
+          either (error . pretty) id $
+            readContract'
               metadataRegistryContractPath
               $(embedStringFile metadataRegistryContractPath)
         |]
   )
+
+----------------------------------------------------------------------------
+-- TZIP-16 Metadata
+----------------------------------------------------------------------------
+
+metadataMap :: MetadataMap BigMap
+metadataMap = BigMap $ Map.fromList
+  -- One might reasonable expect that the URI would be stored as packed Michelson strings,
+  -- but the TZIP-16 spec is explicit about that not being the case.
+  --
+  -- > Unless otherwise-specified, the encoding of the values must be the direct stream
+  -- > of bytes of the data being stored. (...)
+  -- > There is no implicit conversion to Michelson's binary format (PACK) nor
+  -- > quoting mechanism.
+  --
+  -- See: <https://gitlab.com/tzip/tzip/-/blob/eb1da57684599a266334a73babd7ba82dbbbce66/proposals/tzip-16/tzip-16.md#contract-storage>
+  --
+  -- So, instead, we encode it as UTF-8 byte sequences.
```

Comment:
This is something I wondered about while going through the TZIP-16. IIRC It does not mention an encoding here, and just provide some samples. I can't think of a reason why they have left it unspecified.

Reply:
Yeah, I was confused by that too.

The examples they give are examples of ASCII encoding:

> For instance, an URI starting with http: will start with the 5 bytes 0x687474703a (h is 0x68, t is 0x74, etc.).

I think they're assuming URIs can only contain ascii characters (which I don't think is [true](https://en.wikipedia.org/wiki/Internationalized_domain_name)), and so they're implying it should be ASCII-encoded (which is a subset of UTF-8-encoding).

I agree, it should be explicit in the TZIP.

The same goes for the metadata JSON, which definitely can contain non-ascii unicode characters, and therefore could be encoded as UTF-8, UTF-16, ...